076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+   3. During their rolling update step, roll the Kubernetes Pods for ZooKeeper and Kafka to use the new certificates (this is triggered by diffing the server-cert-hash annotation).
+5. The EntityOperator, CruiseControl and KafkaExporter reconcilers generate new certificates signed by the new CA key and roll the Kubernetes Pods to use the new certificates.
+6. A new reconciliation starts.
+7. The CaReconciler updates the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret with a new certificate signed by the new CA certificate (since the Kafka and ZooKeeper Kubernetes Pods now have the strimzi.io/cluster-ca-cert-generation and strimzi.io/cluster-ca-key-generation matching the current Cluster CA and therefore the Cluster CA is fully trusted and used).
+8. The CaReconciler removes the old CA certificates and keys that have been named ca-YYYY-MM-DDTHH-MM-SSZ.crt.
+9. The component reconcilers roll their Kubernetes Pods, since the Cluster CA passed in has a flag indicating certs have been removed.
+
+#### Fig 4: Cluster CA private key update flow
+
+![Cluster CA private key update flow](./images/076-cert-update-existing.png)
+
+> Fig 4: The numbers in the diagram match the numbers in the full description of the steps.
+> The diagram omits the use of annotations, the EntityOperator, CruiseControl and KafkaExporter reconcilers, and the removal of trust of the old CA certificate and key.
+
+The existing logic assumes a few things:
+ - All certificates used by components have been signed by the same root CA certificate.
+ - The CaReconciler performs its reconciliation steps to (if required) either create or update the CA certificates and keys before the other reconcilers reconcile their resources, i.e. the CA certificate is already present when e.g. the KafkaReconciler starts.
+ - There is a single Kubernetes Secret containing the Cluster CA certificate that is used by:
+   - The Strimzi Cluster CA for storing the most recent CA certificate.
+   - The components for mounting into the Kubernetes Pods and used as a Trust Store.
+
+## Motivation
+
+In future it would be beneficial if Strimzi can integrate with commonly deployed certificate management solutions, such as [cert-manager][cmio], or [Vault][vault], rather than managing the Cluster CA itself.
+These solutions work by providing an API that can be used to request certificates from a particular issuer.
+
+Integrating with these solutions requires a change in how things are managed. Specifically:
+ - Strimzi cannot guarantee that the issuer will issue every certificate signed by the same root CA certificate. 
+   - For example cert-manager will if possible include the CA certificate to trust when it issues the certificate (if the CA certificate is not a public one), however this is per certificate and not always provided.
+   - Using those solutions therefore requires TLS peers to be able to trust multiple root CA certificates.
```

Comment:
I still don;t understand why would this be the case. You are still requesting the certificate from a particular issuer and thus a particular CA.

Reply:
The certificate will likely be requested through the `Certificate` resource (https://cert-manager.io/docs/usage/certificate/). And the certificate resource references an issuer. So from my point of view, that issuer identifies the CA. So it is not like you request a certificate and the Cert Manager gives it to you randomly - for broker 1 from Vault and for broker 2 from Venafi. And if some issuer types cannot provide that, maybe we cannot support them.

I think you also need to look at it in the practical terms -> how do you expect it to work when you are in the middle of scale-up or rolling update, you need a new certificate and you suddenly get a new CA to trust? Can you really drop everything and start at the beginning to roll out trust to new CA? I don't think it is that simple.

I think you need to know the CAs upfront in a deterministic way. Whether it has the form of some Cert Manager issuer reference or a Secret reference, I don't think without that assumption you will ever implement it. And I think at the end the clients need it as well. Because they need to know what CAs to trust. If the operator does not know what are the CAs that will be used in one minute, then the client will not know it either and will not know how to connect to the brokers.

Reply:
> And the certificate resource references an issuer. So from my point of view, that issuer identifies the CA.

You're ascribing properties to the issuer that it doesn't necessarily have: That the same issuer always refers to the same thing and that thing corresponds to a unique CA certificate. 

> I think you need to know the CAs upfront in a deterministic way.

Did you mean "CA certificate" there? The problem I worry about a person wanting to a given CA to issue certificates. That person gives that CA a name (the "issuer"). That person cannot know whether that CA will continue to use the same CA certificate forever. The CA can decide to start issuing certificates using some other CA cert at any time. When that happens the person still knows that CA by the same name (the issuer doesn't change), and there's no mechanism for that person to find out that the CA has done this. In other words, having a stable identifier for the CA is not enough to stably identify the CA certificates it will use in signing certificates.

Reply:
> > And the certificate resource references an issuer. So from my point of view, that issuer identifies the CA.
> 
> You're ascribing properties to the issuer that it doesn't necessarily have: That the same issuer always refers to the same thing and that thing corresponds to a unique CA certificate.

Maybe we cannot support such issuers in that case.

But I'm not sure how exactly would that work as you still need to establish trust to make TLS work. So without predictability of who signs the certificates it does not work. At the end, that is how the public CAs work as well - the trusted authorities are what is in the various certificate sets in your OS, in JDK etc.

> > I think you need to know the CAs upfront in a deterministic way.
> 
> Did you mean "CA certificate" there? The problem I worry about a person wanting to a given CA to issue certificates. That person gives that CA a name (the "issuer"). That person cannot know whether that CA will continue to use the same CA certificate forever. The CA can decide to start issuing certificates using some other CA cert at any time. When that happens the person still knows that CA by the same name (the issuer doesn't change), and there's no mechanism for that person to find out that the CA has done this. In other words, having a stable identifier for the CA is not enough to stably identify the CA certificates it will use in signing certificates.

Yeah, sorry. For me CA and CA certificate are the same here. Although your terminology is surely the correct one. So I meant CA certificate. As I said above, I do not think the CA world works on randomness. The idea that the certificate I'm gonna request will be signed by a brand new CA certificate something just decided to generate does not work because nobody will have the trust. So it is not only something that causes problems to us but also to all clients connecting to our clusters.